packages/data/addon/adapters/bard-facts-v2.ts

```diff
@@ -0,0 +1,406 @@
+/**
+ * Copyright 2020, Yahoo Holdings Inc.
+ * Licensed under the terms of the MIT license. See accompanying LICENSE.md file for terms.
+ *
+ * Description: The adapter for the bard-facts request v2 model.
+ */
+
+import { assert } from '@ember/debug';
+import { inject as service } from '@ember/service';
+import { A as array } from '@ember/array';
+import { assign } from '@ember/polyfills';
+import EmberObject from '@ember/object';
+import { canonicalizeMetric, serializeParameters, getAliasedMetrics, canonicalizeAlias } from '../utils/metric';
+import { configHost } from '../utils/adapter';
+
+const SORT_DIRECTIONS = ['desc', 'asc'];
+
+type RequestOptions = {
+  clientId?: string;
+  customHeaders?: Dict<string>;
+  timeout?: number;
+  page?: number;
+  perPage?: number;
+  format?: string;
+  cache?: boolean;
+  queryParams?: Dict<string>;
+};
+type Query = RequestOptions & Dict<string | number | boolean>;
+type SortDirection = typeof SORT_DIRECTIONS[number];
+type ColumnType = 'metric' | 'dimension' | 'timeDimension';
+type Parameters = Dict<string>;
+type Column = {
+  field: string;
+  parameters: Parameters;
+  type: ColumnType;
+  alias: string;
+};
+type Filter = {
+  field: string;
+  parameters: Parameters;
+  type: ColumnType;
+  operator: string;
+  values: string[];
+};
+type Sort = {
+  field: string;
+  parameters: Parameters;
+  direction: SortDirection;
+};
+type Request = {
+  filters: Filter[];
```

Comment:
Are we expecting that any of these might be null/undefined? I see some cases below where we add `?.` for runtime safety so we should probably clean that up either by

- at least have the functions take in `Partial<Request>` to force us to have build and run time safety for partial request objects which is effectively what we're doing now, but manually
- maybe assume that we strictly follow our types and possibly adding some asserts to be safer, but less flexible
- or by only passing `Column[]|undefined`/`Filter[]|undefined`/etc to clarify what the function actually needs

Reply:
I think we shouldn't allow sort or filter to be undefined/null on the request. They should always at least have an empty array. The `?.` checks are probably just leftover from the v1 adapter, but aren't necessary anymore.

Reply:
Ok that's what I was hoping 🙏